BBC-archive/psammead
# Storybook V6 upgrade

Migrating Storybook from v 5.3.19 to 6.X.X

## Resolved issues:

Some minor stuff around:
- Updated integration of a11y addon
- Implementation of new `manager.js` `main.js` `preview.js` structure
- Removed use of `addParameters` from `preview.js`, migrated to the new method
- Importing a `storybook.css` file to overwrite new padding being applied to the story `<body>` tag

### 1. Components aren't re-rendering when navigating around Storybook, or modifying knobs. 
**[FIXED ✅ ]** 
_`6.0.2` & `6.0.21` (with or without forcing `@storybook/core-events` resolutions)._ 

**tl;dr: Fixed by migrating from `notes` to `docs` and removing the `storybook-readme` package. Note that docs is far more 'sophisticated' than notes and is trying to do a lot of dynamic stuff to our readmes, that doesn't work very well. One for the backlog**

Knobs and path changes cause an error, and the component does not re-render. 
Console error: 
```
TypeError: Cannot read property 'debug' of undefined
    at PostmsgTransport.handleEvent
```

It turns out this is being caused by our use of `addon-notes`, which has been deprecated in v6 and replaced by [addon-docs](https://github.com/storybookjs/storybook/tree/next/addons/docs)

A bit of digging showed that we still had an old instance of `@storybook/logger`, which was erroring. This was being used by an old instance of `@storybook/components`, which was being used by... `storybook-readme: ^5.0.8` 

Completely nuking `addon-notes` & `storybook-readme` [has fixed the problem,](https://github.com/bbc/psammead/pull/3781) now to migrate to `addon-docs`! [Migration docs](https://github.com/storybookjs/storybook/blob/next/addons/docs/docs/recipes.md#migrating-from-notesinfo-addons).

---

### 2. Infinite spinner on-load. 
**[FIXED ✅ ]** `6.0.21` 

**tl;dr: This is no longer triggered. 6.0.21 can safely be loaded without an infinite spinner, and without forcing the `"@storybook/core-events": "6.0.2",` resolution.**

---

### 3. Components` required props are undefined
**[FIXED ✅ ]**.  _`6.0.2` & `6.0.21` (with or without forcing `@storybook/core-events` resolutions)._ 

**tl;dr:  'Args' are now the first argument passed into a story, which means our decorators (i.e. withServiceKnob) were not being interpreted correctly as component props. https://github.com/storybookjs/storybook/blob/master/MIGRATION.md#args-passed-as-first-argument-to-story.**

There are two potential fixes for this:

Migrate from 

```js
({ service }) => (
      <Component
         service={service}/>
```

to

```js
( _args, { service }) => (
      <Component
         service={service}/>
```

Or add `passArgsFirst: false,` to `config.js` (which is what I've done)

before: 
<img width="550" alt="Screenshot 2020-09-17 at 16 54 34" src="https://user-images.githubusercontent.com/9042672/93495879-84b81d00-f906-11ea-979a-3ef4bb237813.png">

after: 
<img width="507" alt="Screenshot 2020-09-22 at 10 19 11" src="https://user-images.githubusercontent.com/9042672/93864689-29e34480-fcbd-11ea-97f1-6bd2b81ac8c9.png">

---

### 4. `|` is no longer a valid story hierarchy separator
**[FIXED ✅ ]**  

**tl;dr: replaced `|` with `/`**

We use this throughout. It needs to be replaced by `/`.

**Impact:**
- Component tree visualisation is gubbed.
<img width="204" alt="Screenshot 2020-09-17 at 15 35 21" src="https://user-images.githubusercontent.com/9042672/93496121-c1841400-f906-11ea-9ac3-a52752294766.png">

---


### 5. Arbitrary decorators, and those added after the first story now cause an error
**[FIXED ✅ ]** `6.0.21` 

**tl;dr: removed any arbitrary `addDecorator` funcs in the middle of components**

 i.e. https://github.com/bbc/psammead/blob/2de4871bc90c5be179594b6962ee36037d0f03ff/packages/components/psammead-media-player/src/index.stories.jsx#L130 Although our decoration of stories has tended to be pretty arbitrary, it seems like this component is the only one triggering these errors, so should be a straightforward fix.

**Impact:**
 - Triggers error `You cannot add a decorator after the first story for a kind.` Storybook won't render

---

### Future work:
- **Currently best practice, but soon to be enforced. As outlined [here](https://medium.com/storybookjs/storybook-6-migration-guide-200346241bb5)**
- Do you use storiesOf to write your stories? Upgrade to Component Story Format (CSF). Or choose MDX if you want more control over your documentation.
- Do you use config.js to configure Storybook? Main.js declarative configuration, introduced in 5.3, is a huge improvement and the strong recommendation. Upgrade to main.js.
- Do you use addon-knobs? Storybook Controls is the no-code successor to Knobs that auto-generates the controls based on your component props. Here’s how to migrate your stories to controls.
- Do you use global parameters or decorators? In 6.0 you should add them declaratively; we’ve deprecatedaddParameters/addDecorator.

---

### Testing notes:

- Checkout the `dependabot-storybook` branch locally within Psammead
- Run storybook `npm run storybook`
- Check that there are no issues as you navigate around the components (specifically those above, but keep an eye for any weirdness). Comparing it with https://bbc.github.io/psammead/ will be useful for this

---

- [ ] Automated jest tests added (for new features) or updated (for existing features)
